### PR TITLE
WIP: Add support for testing upgrade scripts

### PR DESCRIPTION
**Pull request by decibel**
Add support for testing of pgTap update scripts.

This commit adds several new make targets:

- make uninstall-all: remove ALL installed pgtap code. Unlike `make unintall`, this removes pgtap*, not just our defined targets. Useful when testing multiple versions of pgtap.
- make regress: run installcheck then print any diffs from expected output.
- make updatecheck: install an older version of pgTap from PGXN (controlled by $UPDATE_FROM; 0.95.0 by default), update to the current version via ALTER EXTENSION, then run installcheck.
- make results: runs `make test` and copies all result files to test/expected/. DO NOT RUN THIS UNLESS YOU'RE CERTAIN ALL YOUR TESTS ARE PASSING!

In addition to these changes, `make installcheck` now runs as many tests as possible in parallel. This is much faster than running them sequentially. The degree of parallelism can be controlled via `$PARALLEL_CONN`. Setting `$PARALLEL_CONN` to `1` will go back to a serial test schedule.

### CODE REVIEW

**Comment by decibel** [comment]
@theory do you want to review this before I commit it?

**Comment by theory** [comment]
> @theory do you want to review this before I commit it?

I'll take a look when I have a few spare minutes. Thanks! Looks like some super useful improvements to managing the code.

File: Makefile [hunk 1]
```diff
@@ -11,10 +11,56 @@ TESTS        = $(wildcard test/sql/*.sql)
...
+# The actual schedule files. Note that we'll build 2 additional files
+SCHEDULE_FILES = $(wildcard test/schedule/*.sch)
+
+# These are our actual regression tests
+TEST_FILES 	= $(filter-out $(SCHEDULE_DEST_FILES),$(wildcard test/sql/*.sql))
+
+# Plain test names
+TESTS		= $(notdir $(TEST_FILES:.sql=))
+
+# Some tests fail when run in parallel
```
[the remark this answers is not in the document]
Comment (decibel) [reply Makefile]: IIRC they were naming conflicts. Ideally we'd fix them. #217

File: Makefile [hunk 2]
```diff
@@ -90,34 +187,38 @@ endif
 
 # Enum tests not supported by 8.2 and earlier.
 ifeq ($(shell echo $(VERSION) | grep -qE "^8[.][12]" && echo yes || echo no),yes)
-TESTS   := $(filter-out test/sql/enumtap.sql,$(TESTS))
-REGRESS := $(filter-out enumtap,$(REGRESS))
+#TESTS   := $(filter-out test/sql/enumtap.sql,$(TESTS))
+#REGRESS := $(filter-out enumtap,$(REGRESS))
```
[the remark this answers is not in the document]
Comment (decibel) [reply Makefile]: Good catch.

File: Makefile [hunk 3]
```diff
@@ -191,6 +292,7 @@ endif
 sql/uninstall_pgtap.sql: sql/pgtap.sql test/setup.sql
 	grep '^CREATE ' sql/pgtap.sql | $(PERL) -e 'for (reverse <STDIN>) { chomp; s/CREATE (OR REPLACE )?/DROP /; print "$$_;\n" }' | sed 's/DROP \(FUNCTION\|VIEW\|TYPE\) /DROP \1 IF EXISTS /' > sql/uninstall_pgtap.sql
 
+# Hmm... is there a race condition here? Will make remove $@ if the recipe dies part-way through?
```
[the remark this answers is not in the document]
Comment (decibel) [reply Makefile]: I've improved things a bit here; still some TODOs.

File: Makefile [hunk 4]
```diff
@@ -214,15 +316,170 @@ sql/pgtap-schema.sql: sql/pgtap-static.sql
...
+# Generated schedule files, one for serial one for parallel
+.PHONY: $(TB_DIR)/tests # Need this target to force schedule rebuild if $(TEST) changes
+$(TB_DIR)/tests: $(TB_DIR)/
+	@[ "`cat $@ 2>/dev/null`" = "$(TEST)" ] || (echo "Rebuilding $@"; echo "$(TEST)" > $@)
+
+.PHONY: $(TB_DIR)/exclude_tests # Need this target to force schedule rebuild if $(EXCLUDE_TEST) changes
+$(TB_DIR)/exclude_tests: $(TB_DIR)/
+	@[ "`cat $@ 2>/dev/null`" = "$(EXCLUDE_TEST)" ] || (echo "Rebuilding $@"; echo "$(EXCLUDE_TEST)" > $@)
+
+$(TB_DIR)/serial.sch: $(GENERATED_SCHEDULE_DEPS)
```
[the remark this answers is not in the document]
Comment (decibel) [reply Makefile]: Lemme know if you wanna hop on a call to walk through it, or just write down a list of questions so I can improve the comments.

File: test/expected/extension.out [hunk 5]
```diff
@@ -12,9 +12,9 @@ ok 9 - extensions_are(exts, desc) should have the proper diagnostics
 ok 10 - extensions_are(exts) should pass
 ok 11 - extensions_are(exts) should have the proper description
 ok 12 - extensions_are(exts) should have the proper diagnostics
-ok 13 - extensions_are(public, exts, desc) should pass
-ok 14 - extensions_are(public, exts, desc) should have the proper description
-ok 15 - extensions_are(public, exts, desc) should have the proper diagnostics
+ok 13 - extensions_are(ci_schema, exts, desc) should pass
+ok 14 - extensions_are(ci_schema, exts, desc) should have the proper description
+ok 15 - extensions_are(ci_schema, exts, desc) should have the proper diagnostics
```
[the remark this answers is not in the document]
Comment (decibel) [reply test/expected/extension.out]: Heh, at this point I don't remember what drove me to do that; I think there was some issue with just plopping everything into public. Actually, that might have been a parallel run issue. In any case, it seems like a good idea for the tests not to assume they're in public.

File: test/sql/extension.sql [hunk 6]
```diff
@@ -15,8 +15,9 @@ BEGIN
     IF pg_version_num() >= 90100 THEN
         EXECUTE $E$
             CREATE SCHEMA someschema;
+            CREATE SCHEMA ci_schema;
```
[the remark this answers is not in the document]
Comment (decibel) [reply test/sql/extension.sql]: Sadly I don't recall...

File: sql/pgtap--0.97.0--0.98.0.sql.in [hunk 7]
```diff
@@ -233,6 +233,23 @@ BEGIN
 END;
 $$ LANGUAGE plpgsql;
 
+-- table_owner_is ( table, user, description )
+CREATE OR REPLACE FUNCTION table_owner_is ( NAME, NAME, TEXT )
```
[the remark this answers is not in the document]
Comment (decibel) [reply sql/pgtap--0.97.0--0.98.0.sql.in]: Yes :(

**Pull request merged**